File: app/src/main/java/se/gustavkarlsson/aurora_notifier/android/dagger/modules/AuroraFactorsModule.java
package se.gustavkarlsson.aurora_notifier.android.dagger.modules;

import dagger.Binds;
import dagger.Module;
import dagger.Reusable;
import se.gustavkarlsson.aurora_notifier.android.background.providers.AuroraFactorsProvider;
import se.gustavkarlsson.aurora_notifier.android.background.providers.impl.aggregating_aurora_factors.AggregatingAuroraFactorsProvider;

@Module
public abstract class AuroraFactorsModule {

	@Binds
	@Reusable
	abstract AuroraFactorsProvider bindAuroraFactorsProvider(AggregatingAuroraFactorsProvider provider);

}

File: app/src/main/kotlin/se/gustavkarlsson/skylight/android/dagger/modules/AuroraFactorsModule.kt
package se.gustavkarlsson.skylight.android.dagger.modules

import dagger.Module
import dagger.Provides
import dagger.Reusable
import io.reactivex.Flowable
import se.gustavkarlsson.skylight.android.entities.*
import se.gustavkarlsson.skylight.android.services.Streamable
import se.gustavkarlsson.skylight.android.services.providers.*
import se.gustavkarlsson.skylight.android.services_impl.providers.CombiningAuroraFactorsProvider
import se.gustavkarlsson.skylight.android.services_impl.streamables.CombiningAuroraFactorsStreamable

@Module
class AuroraFactorsModule {

	@Provides
	@Reusable
	fun provideAuroraFactorsProvider(
		kpIndexProvider: KpIndexProvider,
		visibilityProvider: VisibilityProvider,
		darknessProvider: DarknessProvider,
		geomagLocProvider: GeomagLocationProvider
	): AuroraFactorsProvider = CombiningAuroraFactorsProvider(kpIndexProvider, visibilityProvider, darknessProvider, geomagLocProvider)

	@Provides
	@Reusable
	fun provideAuroraFactorsStreamable(
		kpIndexes: Flowable<KpIndex>,
		visibilities: Flowable<Visibility>,
		darknesses: Flowable<Darkness>,
		geomagLocations: Flowable<GeomagLocation>
	): Streamable<AuroraFactors> = CombiningAuroraFactorsStreamable(kpIndexes, visibilities, darknesses, geomagLocations)

	@Provides
	@Reusable
	fun provideAuroraFactorsFlowable(
		streamable: Streamable<AuroraFactors>
	): Flowable<AuroraFactors> = streamable.stream
}
